elastic/beats
# Refactor of system/memory metricset

Note that this won't work properly before https://github.com/elastic/beats/pull/26251 is merged and I've re-tested everything with that. However we might also need some additional changes, discussed here: https://github.com/elastic/go-structform/pull/32#pullrequestreview-680978188



## What does this PR do?

This is a continuation of the ongoing refactor of the system module, this time for `system/memory`. This is considerably simpler than the changes in `system/cpu`, as the data requires less transformation. 

This is also a first pass at an attempt to move towards more "reporting-defined structures" within the system module, as the main `Memory` struct maps to the final data definition passed upstream. This requires us to do less manual mapping of data, and turns over most of the complexity to the conversion functions. This is designed to work with the `OptX` types introduced in the `system/cpu` refactor, as the above PR to `go-structform` will properly see and omit the `nil` types.

-->

## Why is it important?

This is part of an ongoing refactor of the system code, with a few goals

- centralize metrics-gathering code in one place
- Make OS-specific logic more manageable
- Prevent us from reporting "null zero" metrics 
- (eventually) deprecate `gosigar`

## Checklist


## How to test this PR locally

- Pull down PR and build on any supported platform
- Test the `system/memory` metricset, make sure all fields are reporting properly, and all the data looks reasonable.
